File: adapters/telemetry/selfheal.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Self-healing transport — Merkle-committed blocks + selective retransmission, reusing the repo's real
Merkle primitive (`prismpath.ledger_ots`, which is also OTS/Bitcoin-anchorable). NOT a new subsystem.

(The doc named `audit_log`'s MMR, but that is a stub in the open release — no roots/proofs. `ledger_ots`
is the genuine one. A batch Merkle tree recomputed per window/epoch fits the daily-chained-epoch design;
an append-only MMR with O(log n) peaks is a later streaming optimization.)

The self-framing Fibonacci stream is chunked into fixed-size blocks (the retransmit + Merkle-leaf unit).
The root commits to every block; a lost block is a known gap the receiver requests by index; a
retransmitted (or first-delivery) block is accepted ONLY if its inclusion proof verifies against the
trusted (anchored) root — so a forged or corrupted block is detected, never silently accepted. A block
that is never recoverable stays a *provable* gap: "detect + selectively repair + prove", not magic.
"""
from __future__ import annotations

import hashlib
from typing import Dict, List, Optional, Tuple

from prismpath import ledger_ots as _mk


def chunk(bits: str, block_bits: int) -> List[str]:
    """Split a bitstream into fixed-size blocks (the retransmit unit). Always >= 1 block."""
    if block_bits <= 0:
        raise ValueError("block_bits must be positive")
    return [bits[i:i + block_bits] for i in range(0, len(bits), block_bits)] or [""]


def _leaf(block: str) -> str:
    return hashlib.sha256(block.encode("ascii")).hexdigest()


def commit(blocks: List[str]) -> Tuple[str, List[list]]:
    """(root_hex, per-block inclusion proof) over the block hashes — reuses ledger_ots's Merkle."""
    return _mk.merkle_root_and_paths([_leaf(b) for b in blocks])


def verify_block(block: str, proof: list, root: str) -> bool:
    return _mk.verify_leaf(_leaf(block), proof, root)


class Sender:
    """Holds the committed blocks; serves any block + its proof on a retransmit request."""

    def __init__(self, bits: str, block_bits: int):
        self.blocks = chunk(bits, block_bits)
        self.block_bits = block_bits
        self.root, self.proofs = commit(self.blocks)

    def n_blocks(self) -> int:
        return len(self.blocks)

    def serve(self, idx: int) -> Tuple[str, list]:
        return self.blocks[idx], self.proofs[idx]


class Receiver:
    """Holds only the trusted root (e.g. OTS-anchored) + accepted blocks. Every block — first delivery or
    retransmit — must verify against the root before it is accepted."""

    def __init__(self, root: str, n_blocks: int):
        self.root = root
        self.n = n_blocks
        self._blocks: Dict[int, str] = {}

    def accept(self, idx: int, block: str, proof: list) -> bool:
        """Verify the block's proof against the root; accept iff valid. Returns acceptance."""
        if not (0 <= idx < self.n):
            return False
        if not verify_block(block, proof, self.root):
            return False                       # forged / corrupted -> reject, stays a gap
        self._blocks[idx] = block
        return True

    def missing(self) -> List[int]:
        return [i for i in range(self.n) if i not in self._blocks]

    def complete(self) -> bool:
        return len(self._blocks) == self.n

    def assemble(self) -> str:
        """The recovered bitstream. Raises if any block is still missing — a gap is provable, never a
        silent hole."""
        if not self.complete():
            raise ValueError(f"cannot assemble: blocks still missing (provable gap): {self.missing()}")
        return "".join(self._blocks[i] for i in range(self.n))


def repair(sender: Sender, receiver: Receiver) -> List[int]:
    """Selective retransmission: fetch + verify only the blocks the receiver is missing. Returns the list
    of block indices retransmitted (the selective-repair cost — cf. the Phase A benchmark)."""
    retransmitted = []
    for idx in receiver.missing():
        block, proof = sender.serve(idx)
        if receiver.accept(idx, block, proof):
            retransmitted.append(idx)
    return retransmitted

File: prismpath-telemetry-rs/src/selfheal.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Self-healing transport — Merkle-committed blocks + selective retransmission.

use sha2::{Digest, Sha256};
use std::collections::HashMap;

/// Split a bitstream into fixed-size blocks (the retransmit unit). Always >= 1 block.
pub fn chunk(bits: &str, block_bits: usize) -> Result<Vec<String>, String> {
    if block_bits == 0 {
        return Err("block_bits must be positive".to_string());
    }
    if bits.is_empty() {
        return Ok(vec!["".to_string()]);
    }
    let chars: Vec<char> = bits.chars().collect();
    let mut out = Vec::new();
    for chunk in chars.chunks(block_bits) {
        out.push(chunk.iter().collect());
    }
    Ok(out)
}

fn leaf_hash(block: &str) -> String {
    hex::encode(leaf_bytes(block))
}

fn leaf_bytes(block: &str) -> Vec<u8> {
    let mut hasher = Sha256::new();
    hasher.update(block.as_bytes());
    hasher.finalize().to_vec()
}

/// (root_hex, per-block inclusion proof) over block hashes — reproduces ledger_ots's Merkle tree byte-for-byte.
pub fn commit(blocks: &[String]) -> (String, Vec<Vec<(String, String)>>) {
    if blocks.is_empty() {
        return ("".to_string(), vec![]);
    }
    // Build the leaf layer as raw hash bytes directly — no hex round-trip, no fallible decode.
    let mut layers: Vec<Vec<Vec<u8>>> = vec![blocks.iter().map(|b| leaf_bytes(b)).collect()];

    // `layers` is seeded with one non-empty layer above and only grows, so `.last()` is always Some.
    while layers.last().expect("layers is seeded non-empty").len() > 1 {
        let mut cur = layers.last().expect("layers is seeded non-empty").clone();
        if cur.len() % 2 == 1 {
            let last = cur.last().expect("cur is non-empty (len > 1 above)").clone();
            cur.push(last);
        }
        let mut next_layer = Vec::new();
        for i in (0..cur.len()).step_by(2) {
            let mut hasher = Sha256::new();
            hasher.update(&cur[i]);
            hasher.update(&cur[i + 1]);
            next_layer.push(hasher.finalize().to_vec());
        }
        layers.push(next_layer);
    }

    let root = hex::encode(&layers.last().expect("layers is seeded non-empty")[0]);

    let mut proofs = Vec::new();
    for idx in 0..blocks.len() {
        let mut path = Vec::new();
        let mut cur_idx = idx;
        for layer in layers.iter().take(layers.len() - 1) {
            let mut layer_l = layer.clone();
            if layer_l.len() % 2 == 1 {                    // odd -> duplicate the last (non-empty: len is odd, so >= 1)
                let last = layer_l.last().expect("layer is non-empty (odd length)").clone();
                layer_l.push(last);
            }
            if cur_idx % 2 == 0 {
                path.push(("R".to_string(), hex::encode(&layer_l[cur_idx + 1])));
            } else {
                path.push(("L".to_string(), hex::encode(&layer_l[cur_idx - 1])));
            }
            cur_idx /= 2;
        }
        proofs.push(path);
    }

    (root, proofs)
}

pub fn verify_block(block: &str, proof: &[(String, String)], root: &str) -> bool {
    let leaf_hex = leaf_hash(block);
    let mut cur = match hex::decode(leaf_hex) {
        Ok(c) => c,
        Err(_) => return false,
    };

    for (side, sib) in proof {
        let s = match hex::decode(sib) {
            Ok(bytes) => bytes,
            Err(_) => return false,
        };
        let mut hasher = Sha256::new();
        if side == "R" {
            hasher.update(&cur);
            hasher.update(&s);
        } else {
            hasher.update(&s);
            hasher.update(&cur);
        }
        cur = hasher.finalize().to_vec();
    }

    hex::encode(&cur) == root
}

#[derive(Debug, Clone)]
pub struct Sender {
    pub blocks: Vec<String>,
    pub block_bits: usize,
    pub root: String,
    pub proofs: Vec<Vec<(String, String)>>,
}

impl Sender {
    pub fn new(bits: &str, block_bits: usize) -> Result<Self, String> {
        let blocks = chunk(bits, block_bits)?;
        let (root, proofs) = commit(&blocks);
        Ok(Sender {
            blocks,
            block_bits,
            root,
            proofs,
        })
    }

    pub fn n_blocks(&self) -> usize {
        self.blocks.len()
    }

    pub fn serve(&self, idx: usize) -> (String, Vec<(String, String)>) {
        (self.blocks[idx].clone(), self.proofs[idx].clone())
    }
}

#[derive(Debug, Clone)]
pub struct Receiver {
    pub root: String,
    pub n: usize,
    blocks: HashMap<usize, String>,
}

impl Receiver {
    pub fn new(root: String, n_blocks: usize) -> Self {
        Receiver {
            root,
            n: n_blocks,
            blocks: HashMap::new(),
        }
    }

    pub fn accept(&mut self, idx: usize, block: &str, proof: &[(String, String)]) -> bool {
        if idx >= self.n {
            return false;
        }
        if !verify_block(block, proof, &self.root) {
            return false;
        }
        self.blocks.insert(idx, block.to_string());
        true
    }

    pub fn missing(&self) -> Vec<usize> {
        (0..self.n).filter(|i| !self.blocks.contains_key(i)).collect()
    }

    pub fn complete(&self) -> bool {
        self.blocks.len() == self.n
    }

    pub fn assemble(&self) -> Result<String, String> {
        if !self.complete() {
            return Err(format!(
                "cannot assemble: blocks still missing (provable gap): {:?}",
                self.missing()
            ));
        }
        let mut out = String::new();
        for i in 0..self.n {
            out.push_str(&self.blocks[&i]);
        }
        Ok(out)
    }
}

pub fn repair(sender: &Sender, receiver: &mut Receiver) -> Vec<usize> {
    let mut retransmitted = Vec::new();
    for idx in receiver.missing() {
        let (block, proof) = sender.serve(idx);
        if receiver.accept(idx, &block, &proof) {
            retransmitted.push(idx);
        }
    }
    retransmitted
}
